src/transform-sdk/js/main.cc

```diff
@@ -165,6 +168,35 @@ class record_writer {
         return qjs::value::undefined(ctx);
     }
 
+    std::expected<std::vector<redpanda::header_view>, qjs::exception>
+    extract_headers(JSContext* ctx, const qjs::value& val) {
+        std::vector<redpanda::header_view> headers;
+        if (val.is_undefined() || val.is_null()) {
+            return headers;
+        }
+        if (!val.is_array()) {
+            return std::unexpected(
+              qjs::exception::make(ctx, "unexpected type for headers"));
+        }
+        const size_t len = val.array_length();
+        headers.reserve(len);
+        for (size_t i = 0; i < len; ++i) {
+            auto elem = val.get_element(i);
+            auto key = extract_data(ctx, elem.get_property("key"));
```

Comment:
do we need to check object-ness here?

Reply:
No you just get back undefined, isn't JavaScript great!

<img width="141" alt="Screenshot 2024-04-19 at 4 29 56 PM" src="https://github.com/redpanda-data/redpanda/assets/8669100/14ef3e78-6bda-44bb-b1a9-0c79e7c3ee7a">

Reply:
There was a time in my life where I would put a link to the WAT talk here. Beating a dead horse, etc.

Reply:
Actually I need to check it's not `null` or `undefined`

Reply:
Very familiar with WAT talk. This whole project feels very much like the birth & death of JavaScript: https://www.destroyallsoftware.com/talks/the-birth-and-death-of-javascript

Reply:
Done